webapp/src/blocks.tsx

```diff
@@ -789,68 +802,472 @@ export class Editor extends srceditor.Editor {
+        function getBuiltInBlocks(ns: string) {
+            let cat = snippets.getBuiltinCategory(ns);
+            let blocks: toolbox.BlockDefinition[] = cat.blocks || [];
+            blocks.forEach(b => { b.noNamespace = true })
+            if (!cat.custom && that.nsMap[ns.toLowerCase()]) {
+                blocks = filterBlocks(blocks.concat(that.nsMap[ns.toLowerCase()]));
+            }
+            return blocks;
+        }
+
+        function getNonBuiltInBlocks(ns: string) {
+            return filterBlocks(that.nsMap[ns]) || [];
+        }
+
+        function getExtraBlocks(ns: string, subns: string) {
+            if (subns) return [];
+            let extraBlocks: (toolbox.BlockDefinition | toolbox.ButtonDefinition)[] = [];
+            const onStartNamespace = pxt.appTarget.runtime.onStartNamespace || "loops";
+            if (ns == onStartNamespace) {
+                extraBlocks.push({
+                    name: ts.pxtc.ON_START_TYPE,
+                    attributes: {
+                        blockId: ts.pxtc.ON_START_TYPE,
+                        weight: pxt.appTarget.runtime.onStartWeight || 10
+                    },
+                    blockXml: `<block type="pxt-on-start"></block>`,
+                    noNamespace: true
+                });
+            }
+            // Inject pause until block
+            const pauseUntil = snippets.getPauseUntil();
+            if (pauseUntil && ns == pauseUntil.attributes.blockNamespace) {
+                extraBlocks.push(pauseUntil);
+            }
+            // Add extension buttons
+            if (!subns) {
+                that.extensions.forEach(config => {
+                    const name = config.name;
+                    const namespace = config.extension.namespace || name;
+                    if (ns == namespace) {
+                        extraBlocks.push({
+                            name: `EXT${name}_BUTTON`,
+                            type: "button",
+                            attributes: {
+                                blockId: `EXT${name}_BUTTON`,
+                                label: config.extension.label ? Util.rlf(config.extension.label) : Util.lf("Editor"),
+                                weight: 101
+                            },
+                            callback: () => {
+                                that.openExtension(name);
+                            }
+                        });
+                    }
+                })
+            }
+            return extraBlocks;
+        }
+
+        if (!snippets.isBuiltin(ns)) {
+            return getNonBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        } else {
+            return getBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        }
+    }
+
+    private flyoutBlockXmlCache: pxt.Map<Element[]> = {};
+    public showFlyout(treeRow: toolbox.ToolboxCategory) {
+        const { nameid, subns, icon, color, name, labelLineWidth } = treeRow;
+        const ns = nameid;
```

Comment:
Nitpick: `const { nameid: ns, subns, icon, color, name, labelLineWidth } = treeRow;` gets rid of the extra declaration

Reply:
Cool!